breaking: JSON unescape

xackus opened the pull request:
Unescape JSON strings in the parser.
Looking at the comments in the code suggests it was planned but not done.

Even if we want it as an option, I think escaping should be the default.
I can add the option if there is need.

xackus commented:
Maybe `unescapeStringAlloc` should stay as an internal function. I didn't add bounds checks, because it operated on already validated strings in the parser. Then i thought I could make it available for general use and forgot about it.

xackus commented:
I also just came up with this idea: the parser could count how many bytes will be needed to store the unescaped string and store it in the token. Then you could allocate the correct amount of memory without going through the string another time.

daurnimator commented on lib/std/json.zig:
```diff
@@ -1270,6 +1270,47 @@ pub const Parser = struct {
+/// Unescape a JSON string
+/// Optimized for arena allocators, uses Allocator.shrink
+pub fn unescapeStringAlloc(alloc: *Allocator, input: []const u8) ![]u8 {
+    var output = try alloc.alloc(u8, input.len);
```
This should be `const`

daurnimator commented on lib/std/json.zig:
```diff
@@ -1379,3 +1420,34 @@ test "parsing empty string gives appropriate error" {
+        \\  "formfeed": "\f",
+        \\  "backspace": "\b",
+        \\  "doublequote": "\"",
+        \\  "unicode": "\u0105"
```
Could you add some tests with invalid escapes? e.g. `\x`,  an underlong unicode sequence `\u2abc`

xackus replied:
Invalid escapes are caught earlier in the parser and already tested in json/test.zig

xackus replied:
After checking, there are indeed some cases that are not caught. The tests only ensure that they don't crash the parser.

xackus replied:
Why is `\u2abc` problematic?

xackus replied:
I made the tests for json.StreamingParser run also for json.Parser.
That tests a lot of invalid stuff.

daurnimator replied:
> Why is `\u2abc` problematic?

Doh, I meant to make something up like `\u2xyz`.

xackus commented:
Looks like the parser needs some more work. It only accepts an object as the top level value.

xackus commented:
I added surrogate pair support.
I made the tests for json.StreamingParser run also for json.Parser and made them stricter to verify that correct errors are returned when given invalid surrogate pairs.
This uncovered some bugs, which I fixed (with the exception of an overflow in parseFloat, but that is not a parser bug).

I think this is ready to be merged.

andrewrk commented:
@xackus thanks for this pull request, and apologies for the delay in getting it merged.

daurnimator commented on lib/std/json.zig:
```diff
@@ -898,7 +898,13 @@ pub const TokenStream = struct {
         }
 
-        if (self.parser.complete) {
+        // Without this a bare number fails, becasue the streaming parser doesn't know it ended
```
typo: 'becasue'
